### PR TITLE
Update to use LLVM 7 for compilation

### PR DESCRIPTION
**Pull request by joestringer**
Update container images and test PATH references to use llvm-7 as the
basis for all testing and runtime execution of the compiler.

LLVM-5.0 is picked as the lower bound of valid LLVM versions, because
this version resolved the following issue which the code now relies on:
https://llvm.org/bugs/show_bug.cgi?id=26243

Furthermore, the code generation was improved enough in llvm-5.0 that it
should ease restrictions on code complexity.

That said, we will for now only actively test llvm-7 in the tree so we update
the docs to specify that version as the recommended version.

Related: #5223 

Tasks:
* [x] Once https://github.com/cilium/packer-ci-build/pull/124 is merged, update SERVER_VERSION
* [x] With the new SERVER_VERSION, run both regular and net-next CI runs

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7153)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

EDIT: Green build with old VM. Will update to newer SERVER_VERSION and try that one.

**Review by jrfastab — APPROVED** [review]
LGTM

File: test/runtime/fqdn.go [hunk 1]
```diff
@@ -866,12 +866,8 @@ var _ = Describe("RuntimeFQDNPolicies", func() {
 	})
 
 	Context("toFQDNs populates toCIDRSet when poller is disabled (data from proxy)", func() {
-		var config = `
-PATH=/usr/lib/llvm-3.8/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin
-CILIUM_OPTS=--kvstore consul --kvstore-opt consul.address=127.0.0.1:8500 --debug --pprof=true --log-system-load --tofqdns-enable-poller=false
-INITSYSTEM=SYSTEMD`
 		BeforeAll(func() {
-			vm.SetUpCiliumWithOptions(config)
+			vm.SetUpCiliumWithNoDNSPoller()
```
Comment (eloycoto) [inline test/runtime/fqdn.go]: I would not move this to the helpers. It's easy to know what is doing here if we have all in the test.

Comment (joestringer) [reply test/runtime/fqdn.go]: Fair point.

What I noticed was just that we have three different places where we're specifying full Cilium commandline opts, which means if we change defaults in future then we need to go through each of these and understand what the differences need to be to make a change. IMHO it'd be easier if the three locations applied the changes like "config := defaultConfig + "--to-fqdns-enable-poller=false".

But tidying that up can be subject to a separate PR, I'll remove this refactoring from here.

**Comment by joestringer** [comment]
test-me-please

EDIT: Green runtime build on net-next.

k8s-1.13 doesn't seem to support current net-next kernel, I guess this is expected: 
```
19:17:42     k8s1-1.13: 	[ERROR SystemVerification]: unsupported kernel release: 5.0.0-rc6+
```

**Comment by joestringer** [comment]
test-me-please

EDIT: Seems that loader unit tests failed:

```
	 === RUN   Test
	 START: loader_test.go:143: LoaderTestSuite.TestCompileAndLoad
	 START: loader_test.go:50: LoaderTestSuite.TearDownTest
	 PASS: loader_test.go:50: LoaderTestSuite.TearDownTest	0.001s
	 
	 PASS: loader_test.go:143: LoaderTestSuite.TestCompileAndLoad	0.103s
	 
	 START: loader_test.go:169: LoaderTestSuite.TestCompileFailure
	 level=error msg="Command execution failed" cmd="[llc -march=bpf -mcpu=probe -filetype=obj -o /tmp/cilium_546361365/bpf_lxc.o]" error="context canceled" subsys=datapath-loader
	 START: loader_test.go:50: LoaderTestSuite.TearDownTest
	 PASS: loader_test.go:50: LoaderTestSuite.TearDownTest	0.001s
	 
	 ... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x45BA7A)
	 
	 /usr/local/go/src/runtime/panic.go:513
	   in gopanic
	 /usr/local/go/src/runtime/panic.go:82
	   in panicmem
	 /usr/local/go/src/runtime/signal_unix.go:390
	   in sigpanic
	 compile.go:167
	   in compileAndLink
	 compile.go:230
	   in compile
	 loader.go:73
	   in compileDatapath
	 loader.go:112
	   in compileAndLoad
	 loader_test.go:187
	   in LoaderTestSuite.TestCompileFailure
	 /usr/local/go/src/reflect/value.go:308
	   in Value.Call
	 /usr/local/go/src/runtime/asm_amd64.s:1333
	   in goexit
	 PANIC: loader_test.go:169: LoaderTestSuite.TestCompileFailure
	 
	 START: loader_test.go:152: LoaderTestSuite.TestReload
	 START: loader_test.go:50: LoaderTestSuite.TearDownTest
	 PASS: loader_test.go:50: LoaderTestSuite.TearDownTest	0.001s
	 
	 PASS: loader_test.go:152: LoaderTestSuite.TestReload	0.368s
	 
	 OOPS: 2 passed, 1 PANICKED
	 --- FAIL: Test (0.57s)
```

Split out the fix into https://github.com/cilium/cilium/pull/7170.

Everything else passed.

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/10071/flowGraphTable/

**Comment by joestringer** [comment]
Just waiting for these box builds:
https://jenkins.cilium.io/job/Vagrant-Master-Boxes-Packer-Build/
https://jenkins.cilium.io/job/Vagrant-Master-Boxes-Packer-Build-net-next/

**Comment by joestringer** [comment]
test-me-please

EDIT: Hit #7130:

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/10078/testReport/junit/(root)/runtime/RuntimeFQDNPolicies_Enforces_ToFQDNs_policy/

**Review by borkmann — APPROVED** [review]

File: contrib/packaging/docker/Dockerfile.runtime [hunk 2]
```diff
@@ -9,16 +9,29 @@ apt-get upgrade -y && \
 #
 apt-get install -y --no-install-recommends \
     gpg gpg-agent libelf-dev libmnl-dev libc6-dev-i386 iptables libgcc-5-dev \
-    bash-completion binutils binutils-dev ca-certificates && \
+    bash-completion binutils binutils-dev ca-certificates clang-7 llvm-7 && \
 apt-get purge --auto-remove && \
-apt-get clean
+apt-get clean && \
+#
+# clang-7
```
Comment (ianvernon) [inline contrib/packaging/docker/Dockerfile.runtime:16]: What are the effects of doing this here vs. in runtime-build?

**Review by ianvernon — APPROVED** [review]
🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀

File: contrib/packaging/docker/Dockerfile.runtime (hunk 2 above)
Comment (joestringer) [reply contrib/packaging/docker/Dockerfile.runtime:16]: runtime-build doesn't operate apt-get with updated/upgraded repositories, so this just seemed like the quickest way to get the up to date version installed into the runtime image (ie without having to also change runtime-build to do update/upgrade).

runtime-base is the basis for all tools we can install via apt that we will need in the final image.

runtime-build is for building the dependencies we'll need in the final image, which typically requires basic tools + links to the specific version of software we want to compile, which we then copy over. The apt dependencies there do not strictly need to be the latest versions.

**Comment by joestringer** [comment]
test-me-please

**Comment by jrfastab** [comment]
Fixes: (LLVM3 generates code that fails verifier here, doing a load, add, load instead of a load + offset. Sure we can mess around with the code and try to convince llvm3 to generate "correct" bpf code but this generates "better" code that also passes verifier.)

https://github.com/cilium/cilium/issues/7052

**Review by brb — APPROVED** [review]

**Pull request merged**